pkg/dio/multi.go

```diff
@@ -0,0 +1,259 @@
+type DynamicMultiWriter interface {
+	io.Writer
+	Add(...io.Writer)
+	Remove(io.Writer)
+	Close() error
+}
+
+type multiWriter struct {
+	mutex sync.Mutex
+
+	writers []io.Writer
+}
+
+func (t *multiWriter) Write(p []byte) (n int, err error) {
+	// stash a local copy of the slice as we never want to write twice to a single writer
+	// if remove is called during this flow
+	wTmp := t.writers
+
+	if verbose {
+		log.Debugf("[%p] writing \"%s\" to %d writers", t, string(p), len(t.writers))
+	}
+	// possibly want to add buffering or parallelize this
+	for _, w := range wTmp {
+		n, err = w.Write(p)
+		if err != nil {
+			if err != io.EOF {
+				return
+			}
+
+			// remove the writer
+			log.Debugf("[%p] removing writer due to EOF", t)
+			t.Remove(w)
+		}
+
+		// FIXME: figure out what semantics we need here - currently we may not write to
+		// everything as we abort
+		if n != len(p) {
+			err = io.ErrShortWrite
+			return
+		}
+	}
+	return len(p), nil
+}
+
+func (t *multiWriter) Add(writer ...io.Writer) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	t.writers = append(t.writers, writer...)
+	if verbose {
+		log.Debugf("[%p] added writer - now %d writers", t, len(t.writers))
+	}
+}
+
+// FIXME: provide a mechanism for selectively closing writers
+//  - currently this closes /dev/stdout and logging as well if present
+func (t *multiWriter) Close() error {
+	for _, w := range t.writers {
+		if c, ok := w.(io.Closer); ok {
+			c.Close()
+		}
+	}
+
+	return nil
+}
+
+// TODO: add a ReadFrom for more efficient copy
+
+// Remove doesn't return an error if element isn't found as the end result is
+// identical
+func (t *multiWriter) Remove(writer io.Writer) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+
+	if verbose {
+		log.Debugf("[%p] removing writer - currently %d writers", t, len(t.writers))
+	}
+	for i, w := range t.writers {
+		if w == writer {
+			t.writers = append(t.writers[:i], t.writers[i+1:]...)
+			// using range directly means that we're looping up, so indexes are now
+			// invalid
+			if verbose {
+				log.Debugf("[%p] removed writer - now %d writers", t, len(t.writers))
+			}
+			return
+		}
+	}
+}
+
+// MultiWriter extends io.MultiWriter to allow add/remove of writers dynamically
+// without disrupting existing writing
+func MultiWriter(writers ...io.Writer) DynamicMultiWriter {
+	w := make([]io.Writer, len(writers))
+	copy(w, writers)
+	return &multiWriter{writers: w}
+}
+
+// DynamicMultiReader adds dynamic add/remove to the base multireader behaviour
+type DynamicMultiReader interface {
+	io.Reader
+	Add(...io.Reader)
+	Remove(io.Reader)
+	Close() error
+}
+
+type multiReader struct {
+	mutex sync.Mutex
+	cond  *sync.Cond
+
+	readers []io.Reader
+	err     error
+}
+
+func (t *multiReader) Read(p []byte) (int, error) {
+	n := 0
+	if verbose {
+		defer log.Debugf("[%p] read \"%s\" from %d readers", t, string(p[:n]), len(t.readers))
+	}
+
+	if t.err == io.EOF {
```

Comment:
IMHO truncating is the way to go as we can't know for sure that draining will work (eg; caller could close the channel before calling Close() hence we may end up with a panic if we try to write to a closed channel)